### DISCUSSION TITLE
Option under relax to only tap exactly on time

### DISCUSSION THREAD

**Discussion by Doksha** [ppy/osu#20086] [Ideas] [answered]
the current relax is able to hit a few ms late, a nice feature would be for there to be an option for relax to only tap exactly when the circle needs to be hit. This would allow for better practicing flow aim with relax (currently you can lag quite behind and the taps will just be slower) 
Alternatively have a slider to adjust how far slow you can aim with relax still tapping before relax doesnt tap.

**Comment by MeerukYang** [comment]
Sorry for my topping this discussion thread. But if this could feature, it would be a perfect way to practice aiming.

Relax mod would still click the object if you're late of timing but hover the cursor by. （see in video）

https://github.com/user-attachments/assets/3c4c9165-efbf-4d92-9d4a-f02844ca7bd7

this is no good for practicing aim since if hit at supposed timing it would miss, but it didn't.

**Comment by Number-3434** [comment]
Any updates on this? I am highly looking forward to this feature being implemented, as I would be much happier to use RX for aim practice this way. My main issue is that you could just overshoot late and then let RX automatically hit the note exactly on time for you, which makes it not very good form aim and especially stream practice. In addition, I think having this setting enabled should increase the star difficulty when used on streams because of this reason. (Currently my best approximation is to play DT with OD11).

**Reply by Joehuu** [reply]
Core devs don't usually implement mods or mod settings. It's mostly or always has been external contributors and they are more than welcome to implement this (given the amount of upvotes this has).

**Comment by Number-3434** [comment]
I have implemented a basic version of this. In the video I also added a "Lookahead" setting, but I've removed it from the [pull request](https://github.com/ppy/osu/pull/34350) since it techinically isn't related.

~EDIT: I've found [this pull request](https://github.com/ppy/osu/pull/32487) instead which does better functionality with some support for replays, and additional potential support for pp change. I've kept my original video however since the base functionality is practically the exact same.~

It turns out [my pull request](https://github.com/ppy/osu/pull/34350) actually does the job better for replays (??). I'm redirecting back to mine instead now. (Both mine and @Empika1's pull request have the same functionality live).

https://github.com/user-attachments/assets/e14d0f00-00cf-4e25-a32f-52256c3ea45d

**Reply by Joehuu** [reply, marked as answer]
https://github.com/ppy/osu/pull/34350

**Reply by Taneta-nemausernames** [reply]
Is the "making relax less relaxing"(paraphrase) issue a blocker? Is this feature locked out?

**Reply by Empika1** [reply]
I hope not. I still think this would be a super cool addition to the game, and probably much more widely used than many of the other mods which have been added.

**Reply by Number-3434** [reply]
There is an issue with my implementation that means you can get note-locked out of stacks, since the mod always taps exactly on time regardless of your cursor position, if you miss a note in a stack it won't recover from it.

**Reply by Taneta-nemausernames** [reply]
Is that not desired behaviour, Relax but it always taps perfectly ?

**Reply by Number-3434** [reply]
Yeah, it probably is. But peppy wanted RX to be as relaxing as possible, so I'm not *too* sure whether getting irreversibly note-locked is acceptable for him.

**Reply by Taneta-nemausernames** [reply]
An alternative would be a mod similar to no 100s and 50s on mcosu.

That + da would offer the precise relax the people in this thread want, while keeping relax relaxed. A compromise for both parties. With the down side of adding an additional mod.

**Reply by Number-3434** [reply]
Probably the closest in the game I can currently get is using Difficulty Adjust with Extended OD -> 11, and then pretend any 50s or 100s are misses.

Basically using this mod would be analogous to having infinitely high OD.

**Comment by jonathant09** [comment]
Any update on this? I hope this can be added to the game, at least tapping on time and not 11ms earlier which is drastically different from normal gameplay.

**Reply by cihe13375** [reply]
> tapping on time and not 11ms earlier which is drastically different from normal gameplay.

Not really, because in NM the clicks can be both early and late. If the 11ms thing is removed then relax would only click late but not click early, which would be very weird (to the extent that one would adjust offset to revert that change for better accuracy).

If it were to be implemented I think the OP (only clicking at exactly ±0ms) is the correct path.
